# IERG3300 Introduction to Stochastic Process
---
## Tutorial 1 (Delivered by Ceyuan YANG)
Date: 16th Jan 2020

Materials comes from Tutorial slides by Ceyuan YANG.


### Example 1
In the Dark Ages, Harvard, Dartmouth, and Yale admitted only male students. Assume that, at that time, 80 percent of the sons of Harvard men went to Harvard and the rest went to Yale, 40 percent of the sons of Yale men went to Yale, and the rest split evenly between Harvard and Dartmouth; and of the sons of Dartmouth men, 70 percent went to Dartmouth, 20 percent to Harvard, and 10 percent to Yale. 

(a) Find the probability that the grandson of a man from Harvard went to Harvard. 

(b) Modify the above by assuming that the son of a Harvard man always went to Harvard. Again, find the probability that the grandson of a man from Harvard went to Harvard.


#### Solution
State space $ S=\{H, D, Y\}$. We need to figure out the transition probability matrix.

The $ij$th entry of the matrix ${P}^n$ gives the prob that starting in state $i$ will be in state $j$ after $n$ steps.

In [110]:
import numpy as np
p=np.array([[0.8,0.0,0.2],[0.2,0.7,0.1],[0.3,0.3,0.4]])
print(p)
p_2=p.dot(p)
print(p_2)

[[0.8 0.  0.2]
 [0.2 0.7 0.1]
 [0.3 0.3 0.4]]
[[0.7  0.06 0.24]
 [0.33 0.52 0.15]
 [0.42 0.33 0.25]]


In [111]:
import numpy as np
p=np.array([[1.0,0.0,0.0],[0.2,0.7,0.1],[0.3,0.3,0.4]])
p_2=p.dot(p)
print(p_2)

[[1.   0.   0.  ]
 [0.37 0.52 0.11]
 [0.48 0.33 0.19]]


### Example 2
A certain calculating machine uses only the digits 0 and 1. It is supposed to transmit one of these digits through several stages. However, at every stage, there is a probability p that the digit that enters this stage will be changed when it leaves and a probability q = 1 − p that it won’t. Form a Markov chain to represent the process of transmission by taking as states the digits 0 and 1. What is the matrix of transition probabilities? Now draw a tree and assign probabilities assuming that the process begins in state 0 and moves through two stages of transmission. What is the probability that the machine, after two stages, produces the digit 0 (i.e., the correct digit)?

> *Borrowed from the exercises of  Statistics 150: Stochastic Processes, Department of Statistics, University of California, Berkeley.*

#### Solution
Taking as states the digits 0 and 1 we identify the following Markov chain (by specifying states and transition probabilites):
$$
    \begin{array}{ccc}
        & 0 & 1 \\
        0 & q & p \\
        1 & p & q
    \end{array}\ ,
$$
where $p+q=1$. Thus, the transition matrix is as follows:
$$ 
P= 
\begin{bmatrix}
    q & p \\
    p & q
\end{bmatrix} = 
\begin{bmatrix}
1-p & p \\
p & 1-p
\end{bmatrix} = 
\begin{bmatrix}
q & 1-q \\
1-q & q 
\end{bmatrix}\ .
$$
It is clear that the probability that the machine will produce 0 if it starts with 0 is $p^2+q^2$.

### Example 3
Discuss the topological properties of the graphs of the following Markov chains. Is that irreducible? What is the periodicity of each state?
$$ 
\text{(a).} 
\begin{bmatrix}
0.5 & 0.5 \\
0.5 & 0.5
\end{bmatrix},\ 
\text{(b).}
\begin{bmatrix}
0.5 & 0.5 \\
1 & 0
\end{bmatrix},\ 
\text{(c).}
\begin{bmatrix}
\frac{1}{3} & 0 & \frac{2}{3} \\
0 & 1 & 0 \\
0 & \frac{1}{5} & \frac{4}{5}
\end{bmatrix},\ 
\text{(d).}
\begin{bmatrix}
0 & 1 \\
1 & 0
\end{bmatrix},\ 
\text{(e).}
\begin{bmatrix}
\frac{1}{2} & \frac{1}{2} & 0 \\
0 & \frac{1}{2} & \frac{1}{2} \\
\frac{1}{3} & \frac{1}{3} & \frac{1}{3}
\end{bmatrix}.
$$

#### Solution
Firstly, we can draw the state diagram for the given Markov chain according to its transition matrix.

Irreducible: Judging from the definition
>  A Markov chain is said to be irreducible if it is possible to get to any state from any state.

Periodicity:
> Find the gcd (Greatest Common Divisor) of all integers $t \geq 1 \ s.t. \ P_{i,i}^{(t)} >0$. 

(a) Irreducible.

(b) Irreducible.

(c) Starting from state 2 remains at 2 forever.

(d) Irreducible.

(e) Irreducible.

### Example 4
Smith is in jail and has 3 dollars; he can get out on jail if he has 8 dollars. A guard agrees to make a series of bets with him. If Smith bets A dollars, he wins A dollars with probability 0.4 and loses A dollars with probability 0.6. 

(a) Find the probability that he wins 8 dollars before losing all of his money if he bets 1 dollar each time (timid strategy). 

(b) he bets, each time, as much as possible but not more than necessary to bring his fortune up to 8 dollars (bold strategy). 

(c) Which strategy gives Smith the better chance of getting out of jail?



#### Solution
We model this problem as a Markov chain and draw the diagram of this Markov chain.

*Detailed diagram can be found in the tutorial slide 8 & 10.* [*LINK*(access via CUHK Network)](https://course.ie.cuhk.edu.hk/~ierg3300/IERG3300Tutorial1.pptx) 

(a) Let $\varphi(i)$ be the probability that the chain reaches states 8 before reaching state 0, starting from state $i$. In other words, if $S_j$ is the first $n \geq 0$ such $X_n=j$, 
$$
\varphi(i)=P_i(S_8<S_0)=P(S_8<S_0|X_0=i). 
$$
Then, we can get this system of linear equations
$$
\begin{align*}
\varphi(i)&=0.4\varphi(i+1)+0.6\varphi(i-1),\ i=1,2,3,4,5,6,7 \\
\varphi(0)&=0 \\
\varphi(8)&=1.
\end{align*}
$$
Solving this system of linear equations,

In [139]:
import numpy as np
A=np.array([[1.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0],[-0.6,1.0,-0.4,0.0,0.0,0.0,0.0,0.0,0.0],[0.0,-0.6,1.0,-0.4,0.0,0.0,0.0,0.0,0.0],[0,0,-0.6,1,-0.4,0,0,0,0],[0,0,0,-0.6,1,-0.4,0,0,0],[0,0,0,0,-0.6,1,-0.4,0,0],[0,0,0,0,0,-0.6,1,-0.4,0],[0,0,0,0,0,0,-0.6,1,-0.4],[0,0,0,0,0,0,0,0,1]])
B=np.array([0,0,0,0,0,0,0,0,1])
A_inv=np.linalg.inv(A)
X=np.dot(A_inv,B)
print(X)

[0.         0.02030135 0.05075337 0.0964314  0.16494845 0.26772403
 0.42188739 0.65313243 1.        ]


We find that
$$ 
\begin{align*}
\varphi &=(\varphi(1), \varphi(2), \varphi(3), \varphi(4), \varphi(5), \varphi(6), \varphi(7))\\
        &=(0.0203, 0.0508, 0.0964, 0.1649, 0.2677, 0.4219, 0.6531)
\end{align*}
$$
$\varphi(i)$ is increasing in $i$, which was expected.

(b) Now the bold strategy is used, the Markov chain and the system of linear equations also change accordingly.
$$
\begin{align*}
\varphi(3)&=0.4\varphi(6)+0.6\varphi(0)\\
\varphi(6)&=0.4\varphi(8)+0.6\varphi(4)\\
\varphi(4)&=0.4\varphi(8)+0.6\varphi(0)\\
\varphi(0)&=0\\
\varphi(8)&=1.
\end{align*}
$$
Solving this system of linear equations,

In [140]:
import numpy as np
A=np.array([[1,0,0,0,0],[-0.6,1,0,-0.4,0],[-0.6,0,1,0,-0.4],[0,0,-0.6,1,-0.4],[0,0,0,0,1]])
B=np.array([0,0,0,0,1])
A_inv=np.linalg.inv(A)
X=np.dot(A_inv,B)
print(X)

[0.    0.256 0.4   0.64  1.   ]


We find that 
$$
\begin{align*}
\varphi &= (\varphi(3), \varphi(4), \varphi(6))\\
        &=(0.256, 0.4, 0.64)
\end{align*}\ .
$$
The following part is a simple numerical simulation for the above results.

In [141]:
import numpy
from random import random
from random import seed
cash=3 #starting cash = 3
wintime=0
# bet =1 is for subquestion (a)
bet=1 
# uncomment the following section for subquestion (b)
#if 8-cash>cash:
#   bet=cash
#else:
#   bet=8-cash
for x in range(100000):
    while cash>0:
        result=random()
        if result>0.60:
            cash=cash+bet
        else:
            cash=cash-bet
        if cash >= 8:
            #print("You win!")
            wintime=wintime+1
            #print(wintime)
            break
    cash=3
prob=wintime/100000
print(prob)

0.09725
